### DISCUSSION TITLE
Commit 7dd91a: Corrected FPU tag register init?

### DISCUSSION THREAD

**Discussion by ecm-pushbx** [dosemu2/dosemu2#1930] [General]
I just found during testing of my debugger that the `RN` command showed half of the FPU registers as `0` and the other half as `empty`. Expected is for all of them to be `empty` if no one initialised them.

I found that rebuilding the most recent dosemu2 fixed the problem. The commit that changes this behaviour is https://github.com/dosemu2/dosemu2/commit/7dd91a04c23adf20ec99c408ac4706ae14e887b2

I might get back to this and test in more detail whether this spot should initialise the tag register to anything, or not. Writing 5555h there certainly wasn't correct, as determined by the debugger's behaviour.

A test program is at https://pushbx.org/ecm/test/20230302/ (Public Domain source + executable), along with two data files as resulting from running dosemu revision 7dd91a and its parent commit. This is the command I used to run the test:

`dosemu -K "$PWD" -E testfpur.com && podhex testfpur.dat && cp -a testfpur.dat 7dd91a.dat`

`podhex` is a bash shell function that is defined as follows:

`function podhex() { od -Ax -tx1z "$@" | perl -pe 's/^([0-9A-Fa-f]+(:? +([0-9A-Fa-f]+ +)+)?)/uc($1)/eg;s/^([0-9A-F]+(?: +[0-9A-F]+){8}) /$1-/g'; }`

The test shows that the tag register is 0DDDDh for the wrong case and 0FFFFh for the correct case. This is the only difference between these files.

**Comment by bartoldeman** [comment]
yes I need to look at that. The problem is that it uses the fnsave tag format for what's put into the fxsave tag in the KVM update.
There are two different kinds of FPU reset, one is what's done by FNINIT, which sets the tag to 0xffff (old) / 0x00 (new), i.e. all empty.
The other is here, and is like an FPU "reboot", 0x5555 means all zero, for FXSAVE that's equivalent to using 0xff for the tag, and 0.0 for all FPU registers.
https://www.felixcloutier.com/x86/fxsave
I've got a long PR open about this going though a lot of iterations, should fix that one too (#1886), it actually went into two related but different issues (consistent save/restore and FP exception handlers), so it should be split in two anyway.
